src/type_icons.c

```diff
@@ -0,0 +1,531 @@
+    LoadCompressedSpritePalette(&sTypeIconPal1);
+    LoadCompressedSpritePalette(&sTypeIconPal2);
+}
+
+static void LoadTypeIconsPerBattler(u32 battler, u32 position)
+{
+    u32 typeNum, types[2];
+    u32 battlerId = GetBattlerAtPosition(position);
+    bool32 useDoubleBattleCoords = UseDoubleBattleCoords(battlerId);
+
+    if (IsBattlerFainted(battlerId))
+        return;
+
+    for (typeNum = 0; typeNum < 2; ++typeNum)
+        types[typeNum] = GetMonPublicType(battlerId, typeNum);
+
+    for (typeNum = 0; typeNum < 2; ++typeNum)
+        CreateSpriteFromType(position, useDoubleBattleCoords, types, typeNum, battler);
+}
+
+static bool32 UseDoubleBattleCoords(u32 position)
+{
+    if (!BATTLE_TYPE_IS_DOUBLE)
+        return FALSE;
+
+    if ((position == B_POSITION_PLAYER_LEFT) && (IsBattlerNull(B_POSITION_PLAYER_RIGHT)))
+        return FALSE;
+
+    if ((position == B_POSITION_OPPONENT_LEFT) && (IsBattlerNull(B_POSITION_OPPONENT_RIGHT)))
+        return FALSE;
+
+    return TRUE;
+}
+
+static bool32 IsBattlerNull(u32 battlerId)
+{
+    return gBattleMons[battlerId].species== SPECIES_NONE;
+}
+
+static bool32 IsBattlerFainted(u32 battlerId)
+{
+    return gBattleMons[battlerId].hp < 1;
+}
+
+static u32 GetMonPublicType(u32 battlerId, u32 typeNum)
+{
+    struct Pokemon* mon = GetBattlerData(battlerId);
+    struct Pokemon* monIllusion = GetIllusionMonPtr(battlerId);
+    u32 illusionSpecies = GetMonData(monIllusion,MON_DATA_SPECIES,NULL);
+    u32 monSpecies = GetMonData(mon,MON_DATA_SPECIES,NULL);
+
+    if (ShouldHideUncaughtType(monSpecies))
+        return TYPE_MYSTERY;
+
+    if (GetActiveGimmick(battlerId) == GIMMICK_TERA)
+        return ReturnMonDefensiveTeraType(mon,monIllusion,battlerId,typeNum,illusionSpecies,monSpecies);
+
+    if (IsIllusionActiveAndTypeUnchanged(monIllusion,monSpecies, battlerId))
+        return gSpeciesInfo[illusionSpecies].types[typeNum];
+
+	return gBattleMons[battlerId].types[typeNum];
+}
+
+static struct Pokemon* GetBattlerData(u32 battlerId)
+{
+    u32 index = gBattlerPartyIndexes[battlerId];
+
+    return (GetBattlerSide(battlerId) == B_SIDE_OPPONENT) ? &gEnemyParty[index] : &gPlayerParty[index];
+}
+
+static bool32 ShouldHideUncaughtType(u32 species)
+{
+    if (B_SHOW_TYPES != SHOW_TYPES_CAUGHT)
+        return FALSE;
+
+    if (GetSetPokedexFlag(SpeciesToNationalPokedexNum(species),FLAG_GET_CAUGHT))
+        return FALSE;
+
+    return TRUE;
+}
+
+static u32 ReturnMonDefensiveTeraType(struct Pokemon * mon, struct Pokemon* monIllusion, u32 battlerId, u32 typeNum, u32 illusionSpecies, u32 monSpecies)
+{
+    u32 teraType = GetBattlerTeraType(battlerId);
+    u32 targetSpecies;
+
+    if (teraType != TYPE_STELLAR)
+        return teraType;
+
+    targetSpecies = (IsIllusionActive(monIllusion)) ? illusionSpecies : monSpecies;
+
+    return gSpeciesInfo[targetSpecies].types[typeNum];
+}
+
+static bool32 IsIllusionActive(struct Pokemon* monIllusion)
+{
+    return (monIllusion != NULL);
```

Comment:
I think this comparison is understandable enough where the function is not needed.

Reply:
https://github.com/rh-hideout/pokeemerald-expansion/pull/5131/commits/747fc3271839a039c895beef283d39deff3bbdca